growth-program/src/rewards/claim.rs

```diff
@@ -194,6 +147,104 @@ pub trait ClaimRewardsModule:
+        {
+            return;
+        }
+
+        self.require_project_active(args.project_id);
+        require!(
+            args.opt_note_and_signature.is_some(),
+            "Must provide note and signature"
+        );
+
+        let (note, signature) = unsafe {
+            args.opt_note_and_signature
+                .into_option()
+                .unwrap_unchecked()
+                .into_tuple()
+        };
+        let signature_data = SignatureData {
+            caller: args.caller,
+            project_id: args.project_id,
+            week: args.current_week,
+            note: &note,
+        };
+
+        self.verify_signature(signature_data, &signature);
+
+        let full_note = Note {
+            note_data: note,
+            week: args.current_week,
+        };
+        self.insert_note(args.project_id, args.user_id, &full_note);
+    }
+
+    fn claim_exemption(&self, args: ClaimExemptionArgs<Self::Api>) {
+        let beta = self.beta().get();
+        let remaining_rewards = self
+            .rewards_remaining_amount(args.project_id, args.current_week)
+            .get();
+
+        require!(
+            remaining_rewards < args.rew_advertised,
+            "Can claim full rewards"
+        );
+        require!(args.min_rewards == 0, "Invalid min rewards");
+
+        self.registered_energy_exemption_claimers(args.project_id, args.current_week)
+            .update(|reg_energy| *reg_energy += args.user_original_energy);
+
+        let rew_available = beta * args.rew_advertised / MAX_PERCENTAGE;
+        let rew_available_dollar_value =
+            self.get_dollar_value(args.reward_token_id, rew_available, HOUR_IN_SECONDS);
+        self.registered_rewards_dollars(args.project_id, args.current_week)
+            .update(|reg_rew_dollars| *reg_rew_dollars += rew_available_dollar_value);
+
+        let _ = self
+            .exempted_participants(args.project_id, args.current_week + 1)
+            .insert(args.user_id);
+    }
+
+    fn claim_rewards_normal(&self, args: ClaimRewardsArgs<Self::Api>) -> EsdtTokenPayment {
+        let rem_rewards_mapper = self.rewards_remaining_amount(args.project_id, args.current_week);
+        let remaining_rewards = rem_rewards_mapper.get();
+
+        let rew_available = core::cmp::min(args.rew_advertised, remaining_rewards);
+        let user_rewards =
+            self.adjust_amount_to_lock_option(rew_available.clone(), args.lock_option);
+        require!(user_rewards >= args.min_rewards, "Too few rewards");
+
+        rem_rewards_mapper.update(|rem_rew| *rem_rew -= &user_rewards);
+
+        self.registered_energy_rewards_claimers(args.project_id, args.current_week)
+            .update(|reg_energy| *reg_energy += &args.user_original_energy);
+
+        let user_adjusted_energy =
+            self.adjust_amount_to_lock_option(args.user_original_energy, args.lock_option);
+        self.interested_energy_rewards_claimers(args.project_id, args.current_week)
```

Comment:
Why do we need 2 storages?
`registered_energy_rewards_claimers` and `interested_energy_rewards_claimers`

Reply:
Those were the specs, you'd have to ask the arda guys.